### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by Alejandro-Andreo**
Bastante parecido al modulo 3. Esta vez, he pulido los archivos que no tienen que ver con el notebook. Aunque esta vez no he hecho tests por falta de tiempo.

Como se ve en la imagen de abajo, el Gradient Boosting Tree es el mejor modelo, Particularmente, he utilizado la libreria de **HistGradientBoostingClassifier** pues se supone que es ordenes de magnitud más rápida que **GradientBoostingClassifier** a la hora de entrenar modelos con tanta cantidad de datos como con el dataset que estamos utilizando. La única pega es que no tiene el atributo de `feature_importances_`, por lo que para averiguarlas he tenido que utilizar el método de la libreria `sklearn.inspection` `permutation_importance()` que tardaba tanto como el propio entrenamiento de un modelo.
Por otro lado, a diferencia de lo que se mostraba en clase, a mí, el RF no me mostraba AUC de 0.85 cuando lo entrenaba únicamente con las features más importantes, ni tampoco me mejoraba el modelo. He estado utilizando random_seed, así que supongo que será por eso.
image

### CODE REVIEW

File: src/module_3/groceries_model.ipynb
Comment (pabaldonedo) [inline src/module_3/groceries_model.ipynb:1]: Branches should come from develop so here no data from module_3 should be present

File: src/module_3/groceries_model_files/groceries_model_10_1.png
Comment (pabaldonedo) [inline src/module_3/groceries_model_files/groceries_model_10_1.png:1]: images should not go to the repo

File: src/module_4/groceries_model_fitting.md [hunk 1]
```diff
@@ -0,0 +1,1642 @@
...
+from sklearn.preprocessing import StandardScaler
+from pathlib import Path
+from IPython.display import display
+```
+
+
+```python
+DATA_PATH = Path("../../..").resolve()
+
+df = pd.read_csv(Path(DATA_PATH, "feature_frame.csv"))
```
Comment (pabaldonedo) [inline src/module_4/groceries_model_fitting.md:37]: You can join paths with the "/" operator

File: src/module_4/groceries_model_fitting.md [hunk 2]
```diff
@@ -0,0 +1,1642 @@
...
+
+    Baseline model: Train AUC = 0.787, Val AUC = 0.787; Train AP = 0.060, Val AP = 0.066; Train Crossentropy = 0.070, Val Crossentropy = 0.064
+
+
+
+    
+![png](groceries_model_fitting_files/groceries_model_fitting_15_1.png)
+    
+
+
```
Comment (pabaldonedo) [inline src/module_4/groceries_model_fitting.md:239]: Up to here I assumed that it was the same as module 3, thus did not check thoroughly and all the comments on module 3 applies

File: src/module_4/groceries_model_fitting.md [hunk 3]
```diff
@@ -0,0 +1,1642 @@
...
+
+    
+![png](groceries_model_fitting_files/groceries_model_fitting_18_4.png)
+    
+
+
+- RF with better peformance in all metrics(roc_AUC, PR_AUC, AP and CrossEntropy) are the ones with more trees.
```
Comment (pabaldonedo) [inline src/module_4/groceries_model_fitting.md:353]: As the theory suggests, so nice.

The cross entropy does not seem to have converged so maybe we can have checked with further trees

File: src/module_4/groceries_model_fitting.md [hunk 4]
```diff
@@ -0,0 +1,1642 @@
...
+        "features": train_cols,
+        "importances_mean": result.importances_mean,
+        "importances_std": result.importances_std,
+    }
+).sort_values("importances_mean", ascending=False)
+```
+
+
+```python
+import seaborn as sns
```
Comment (pabaldonedo) [inline src/module_4/groceries_model_fitting.md:1472]: better to import on top

File: src/module_4/groceries_model_fitting.md
Comment (pabaldonedo) [inline src/module_4/groceries_model_fitting.md:1]: Really solid work, Honestly, little feedback I can give you on the work you have done. REally nice!

The only would be to explore things like categorical features or models like catboost that use them as built in.

File: src/module_4/utils.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+
+def load_dataset() -> pd.DataFrame:
+    """
+    Load the dataset from the data folder.
+    Returns:
+        pd.DataFrame: The input dataframe containing the dataset.
```
Comment (pabaldonedo) [inline src/module_4/utils.py:17]: it would be nice to know which columns it has

File: src/module_4/utils.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import logging
+import pandas as pd
+from pathlib import Path
+
+
+DATA_PATH = Path(__file__).parent.parent.parent.parent
+
+logging.basicConfig(
```
Comment (pabaldonedo) [inline src/module_4/utils.py:8]: Nice use of logging here :D

File: src/module_4/utils.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    Returns:
+        pd.DataFrame: The input dataframe containing the dataset.
+    """
+    logging.info("Loading filtered dataset")
+    return (
+        load_dataset()
+        .pipe(push_relevant_dataframe)
+        .assign(created_at=lambda x: pd.to_datetime(x.created_at))
+        .assign(order_date=lambda x: pd.to_datetime(x.order_date).dt.date)
+    )
```
Comment (pabaldonedo) [inline src/module_4/utils.py:46]: Add final blank line to follow PEP8

File: src/module_4/push_model.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+from sklearn.ensemble import HistGradientBoostingClassifier
+
+
+class Push_Model:
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:4]: Use SnakeCase PushModel to comply with PEP8

File: src/module_4/push_model.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        "std_days_to_buy_product_type",
+        "days_since_purchase_product_type",
+        "avg_days_to_buy_product_type",
+        "user_order_seq",
+        "avg_days_to_buy_variant_id",
+        "set_as_regular",
+    ]
+    TARGET_COL = "outcome"
+
+    def __init__(self, params: dict):
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:18]: Better to use Dict class (from typing import Dict)

Comment (pabaldonedo) [reply src/module_4/push_model.py:18]: Also missing docstring

File: src/module_4/push_model.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        self.model = HistGradientBoostingClassifier(**params)
+        self.best_params = params
+
+    def fit(self, df):
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:22]: type hinting

File: src/module_4/fit.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+from sklearn.base import BaseEstimator
+from push_model import Push_Model
+from utils import load_filtered_data
+
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+
+def save_model(model: BaseEstimator, model_name: str) -> str:
```
Comment (pabaldonedo) [inline src/module_4/fit.py:15]: Type hinting is wrong, it is not BaseESimtator

File: src/module_4/fit.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    Save the model to the models folder.
+
+    Args:
+        model (BaseEstimator): The trained model.
+        model_name (str): The name of the model.
+
+    Returns:
+        str: The path to the saved model.
+    """
+    model_path = Path(Path(__file__).parent, "models", f"{model_name}.pkl")
```
Comment (pabaldonedo) [inline src/module_4/fit.py:26]: Allow model_path as parameter for more flexibility and set this as default. IN additon if you prefer you can join paths or paths with strings with "/"

File: src/module_4/fit.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    pickle.dump(model, open(model_path, "wb"))
```
Comment (pabaldonedo) [inline src/module_4/fit.py:27]: use context manager with open(). You are opening the file without closing it which can lead to corrupting the file

File: src/module_4/fit.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+    Returns:
+        str: The model name.
+    """
+    train_date = datetime.today().strftime("%Y-%m-%d")
+    model_name = "push_" + train_date
+    return model_name
+
+
+def extract_params(event: dict) -> dict:
```
Comment (pabaldonedo) [inline src/module_4/fit.py:47]: use Dict from typing

Comment (pabaldonedo) [reply src/module_4/fit.py:47]: Doctring with expected fields in event

File: src/module_4/inference.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+def load_data(data: dict) -> pd.DataFrame:
+    df = pd.DataFrame.from_dict(data)
+    return load_filtered_data(df)
+
+
+def load_model(model_path: str) -> Push_Model:
+    if not os.path.exists(model_path):
+        raise FileNotFoundError(f"Model not found at {model_path}")
+    model = pickle.load(open(model_path, "rb"))
```
Comment (pabaldonedo) [inline src/module_4/inference.py:17]: context manager with open()

**Review by pabaldonedo — APPROVED** [review]
Super work! I hardly comment much since the work is very solid. Only some minor things on typing , linting etc. In this last thing you can consider use black for style and flake for linting.

What you may be missing you also commented (tests) or expanding (exploring categorical features); so it is fine due to limited time.

Really really good!
